src/ts/ConvivaAnalytics.ts

```diff
@@ -186,89 +218,73 @@ export class ConvivaAnalytics {
     return this.sessionKey !== Conviva.Client.NO_SESSION_KEY;
   }
 
-  private onSourceLoaded = (event: any) => {
+  private onSourceLoaded = () => {
     if (this.isAd) {
       // Ignore ON_SOURCE_LOADED events during ad playback, because that's just an ad being temporarily loaded
       // instead of the actual source.
       return;
     }
 
-    if (this.isValidSession()) {
-      // Do not start a new session when a session is already existing
-      // Happens after ad playback, when the actual source is restored and an ON_SOURCE_LOADED event issued. Because
-      // we suppress the ON_SOURCE_UNLOADED event which unloads the temporary ad source, we must also ignore this
-      // event. By ignoring these ad-induced events, we end up with a clean ON_SOURCE_LOADED/ON_SOURCE_UNLOADED
-      // sequence which only concerns the actual source.
-      return;
-    }
+    // Do not start a new session when a session is already existing
+    // Happens after ad playback, when the actual source is restored and an ON_SOURCE_LOADED event issued. Because
+    // we suppress the ON_SOURCE_UNLOADED event which unloads the temporary ad source, we must also ignore this
+    // event. By ignoring these ad-induced events, we end up with a clean ON_SOURCE_LOADED/ON_SOURCE_UNLOADED
+    // sequence which only concerns the actual source.
+    this.updateSession();
   };
 
   private onReady = (event: any) => {
     this.debugLog('ready', event);
-
-    let config = this.player.getConfig();
-    let autoplayEnabled = config && config.playback && config.playback.autoplay;
-
-    // Start session immediately when autoplay is enabled
-    if (autoplayEnabled) {
-      // Trigger onPlay to create a session similarly to when a user starts playback
-      this.onPlay(event);
-    }
   };
 
   private onPlaybackStateChanged = (event?: any) => {
     this.debugLog('reportplaybackstate', event);
-    let playerState = Conviva.PlayerStateManager.PlayerState.UNKNOWN;
+    let playerState;
 
-    if ((!this.player.isPlaying() && !this.player.isPaused()) || this.player.hasEnded()) {
-      // Before playback has started, and after it is finished, we report the stopped state
-      playerState = Conviva.PlayerStateManager.PlayerState.STOPPED;
```

Comment:
Why don't you set the stopped state from here, but all other states you do?

Reply:
Because I set the stopped state explicitly in the ON_PLAYBACK_FINISHED handler.

Reply:
I see but I think it's much cleaner if one and the same matter is always handled in one and the same place.